From db5399d6552addace841a73320220a2fcee3d3f3 Mon Sep 17 00:00:00 2001
From: Billy Laws <blaws05@gmail.com>
Date: Sat, 16 Nov 2019 17:44:56 -0600
Subject: [PATCH] Add Nvidia custom layer support [2/2]

Change-Id: Ic93e65c15eedc01a12a960f2669cf7aaaedd1c56
---

diff --git a/include/hardware/hwcomposer2.h b/include/hardware/hwcomposer2.h
index c70aef6..8cbb185 100644
--- a/include/hardware/hwcomposer2.h
+++ b/include/hardware/hwcomposer2.h
@@ -290,6 +290,8 @@
      * layer would be. The client may ignore this request if the layer must be
      * blended. */
     HWC2_LAYER_REQUEST_CLEAR_CLIENT_TARGET = 1 << 0,
+    HWC2_LAYER_REQUEST_HINT_COMPOSITION_DEVICE_OVERLAY = 1 << 1,
+    HWC2_LAYER_REQUEST_HINT_LOW_LATENCY = 1 << 2,
 } hwc2_layer_request_t;
 
 /* Power modes for use with setPowerMode */
@@ -610,6 +612,8 @@
     switch (__BIONIC_CAST(static_cast, int, request)) {
         case 0: return "None";
         case HWC2_LAYER_REQUEST_CLEAR_CLIENT_TARGET: return "ClearClientTarget";
+        case HWC2_LAYER_REQUEST_HINT_COMPOSITION_DEVICE_OVERLAY: return "HintCompositionDeviceOverlay";
+        case HWC2_LAYER_REQUEST_HINT_LOW_LATENCY: return "HintLowLatency";
         default: return "Unknown";
     }
 }
@@ -851,6 +855,8 @@
 
 enum class LayerRequest : int32_t {
     ClearClientTarget = HWC2_LAYER_REQUEST_CLEAR_CLIENT_TARGET,
+    HintCompositionDeviceOverlay = HWC2_LAYER_REQUEST_HINT_COMPOSITION_DEVICE_OVERLAY,
+    HintLowLatency = HWC2_LAYER_REQUEST_HINT_LOW_LATENCY,
 };
 TO_STRING(hwc2_layer_request_t, LayerRequest, getLayerRequestName)
 
